### PR TITLE
fix(router): Convert ephemeral to client secret auth list payment_method_customer

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

convert ephemeral key auth to client secret auth, and fetch customer_id from payment_intent using client_secret. Inject customer_id into the payment_method_list flow.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

tested in local
Screen Shot 2023-07-13 at 2 20 56 PM
Screen Shot 2023-07-13 at 2 22 52 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/compatibility/stripe/app.rs [hunk 1]
```diff
@@ -88,7 +88,7 @@ impl Customers {
                     .route(web::delete().to(customer_delete)),
             )
             .service(
-                web::resource("/{customer_id}/payment_methods")
+                web::resource("/payment_methods")
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/app.rs:91]: Why was this changed. It doesn't match stripe endpoint now https://stripe.com/docs/api/payment_methods/customer_list

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]
Why is customer_id being removed ? This API is an endpoint for fetching a customer's payment_methods. You can create a new api endpoint which calls this internally.

File: crates/router/src/compatibility/stripe/app.rs (hunk 1 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/app.rs]: done, fixed it!

File: crates/router/src/compatibility/stripe/customers.rs [hunk 2]
```diff
@@ -167,11 +167,11 @@ pub async fn customer_delete(
 pub async fn list_customer_payment_method_api(
     state: web::Data<routes::AppState>,
     req: HttpRequest,
-    path: web::Path<String>,
+    json_payload: web::Query<payment_methods::PaymentMethodListRequest>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/customers.rs:170]: you need to use the customer_id here. 
For this PR use case create separate API endpoint don't change existing one.

Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/customers.rs:170]: Please don't change existing endpoint. It is meant to be stripe compatible https://stripe.com/docs/api/payment_methods/customer_list

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]
Please create separate endpoint for this use case. Don't change existing stripe compatible endpoint.

File: crates/router/src/compatibility/stripe/customers.rs (hunk 2 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/customers.rs:170]: sure will change this

File: crates/router/src/routes/payment_methods.rs [hunk 3]
```diff
@@ -115,38 +115,44 @@ pub async fn list_payment_method_api(
...
     json_payload: web::Query<payment_methods::PaymentMethodListRequest>,
 ) -> HttpResponse {
     let flow = Flow::CustomerPaymentMethodsList;
-    let customer_id = customer_id.into_inner().0;
-
-    let auth_type = match auth::is_ephemeral_auth(req.headers(), &*state.store, &customer_id).await
-    {
-        Ok(auth_type) => auth_type,
-        Err(err) => return api::log_and_return_error_response(err),
+    let payload = json_payload.into_inner();
```
Comment (Abhicodes-crypto) [inline crates/router/src/routes/payment_methods.rs:129]:
```suggestion
    let payload = query_payload.into_inner();
```

File: crates/router/src/routes/payment_methods.rs [hunk 4]
```diff
@@ -115,38 +115,44 @@ pub async fn list_payment_method_api(
...
     operation_id = "List all Payment Methods for a Customer",
-    security(("api_key" = []), ("ephemeral_key" = []))
+    security(("api_key" = []), ("publishable_key" = []))
 )]
 #[instrument(skip_all, fields(flow = ?Flow::CustomerPaymentMethodsList))]
+#[instrument(skip_all, fields(flow = ?Flow::CustomerPaymentMethodsList))]
 pub async fn list_customer_payment_method_api(
     state: web::Data<AppState>,
-    customer_id: web::Path<(String,)>,
+    customer_id: Option<web::Path<(String,)>>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/routes/payment_methods.rs:124]: Just a ques , what will happen if we write `web::Path<(Option<String>,)>`

File: crates/router/src/routes/app.rs [hunk 5]
```diff
@@ -221,15 +221,19 @@ impl Customers {
         {
             route = route
                 .service(web::resource("").route(web::post().to(customers_create)))
+                .service(
+                    web::resource("/payment_methods")
+                        .route(web::get().to(list_customer_payment_method_api)),
+                )
+                .service(
+                    web::resource("{customer_id}/payment_methods")
```
Comment (Abhicodes-crypto) [inline crates/router/src/routes/app.rs:229]: Shouldn't it be starting with `/` here, Is this flow tested

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/app.rs (hunk 5 above)
Comment (sahkal) [reply crates/router/src/routes/app.rs:229]: yes, but still it was working. have posed ss for the same !

File: crates/router/src/routes/payment_methods.rs (hunk 4 above)
Comment (sahkal) [reply crates/router/src/routes/payment_methods.rs:124]: then the sdk or the client would require to customer_id. where as the client will pass publishable key along with client-secret and we will fetch customer_id based on the client_secret

**Review by Abhicodes-crypto — DISMISSED** [review]

**Review by Abhicodes-crypto — DISMISSED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 6]
```diff
@@ -1499,15 +1499,48 @@ async fn filter_payment_mandate_based(
...
+    key_store: domain::MerchantKeyStore,
+    req: Option<api::PaymentMethodListRequest>,
+    customer_id: Option<&str>,
+) -> errors::RouterResponse<api::CustomerPaymentMethodsListResponse> {
+    let db = &*state.store;
+    if customer_id.is_some() {
+        let customer_id = customer_id.ok_or(errors::ApiErrorResponse::CustomerNotFound)?;
+        list_customer_payment_method(state, merchant_account, key_store, None, customer_id).await
+    } else {
+        let cloned_secret = req.and_then(|r| r.client_secret.as_ref().cloned());
```
Comment (Narayanbhat166) [inline crates/router/src/core/payment_methods/cards.rs:1514]: We are using only `client_secret` from the request, what about other fields in `PaymentMethodsListRequest` ? Don't we need them? If not we can remove them from the struct
https://github.com/juspay/hyperswitch/blob/6447b04574e941b9214239bf5b65b7c1a229dfd6/crates/api_models/src/payment_methods.rs#L324
cc: @bernard-eugine

File: crates/router/src/routes/payment_methods.rs [hunk 7]
```diff
@@ -115,39 +115,92 @@ pub async fn list_payment_method_api(
...
-    let auth_type = match auth::is_ephemeral_auth(req.headers(), &*state.store, &customer_id).await
-    {
-        Ok(auth_type) => auth_type,
-        Err(err) => return api::log_and_return_error_response(err),
+/// List payment methods for a Customer
+///
+/// To filter and list the applicable payment methods for a particular Customer ID
+#[utoipa::path(
+    get,
+    path = "/customers/payment_methods?{client-secret}",
```
Comment (Narayanbhat166) [inline crates/router/src/routes/payment_methods.rs:158]: `client-secret` should not be in the path, add it in params

File: config/development.toml [hunk 8]
```diff
@@ -89,7 +90,6 @@ cards = [
     "shift4",
     "stripe",
     "trustpay",
-    "tsys",
```
[the remark this answers is not in the document]
Comment (sahkal) [reply config/development.toml:92]: not sure, this came from previous main i guess. checked current main adding it back

File: config/development.toml [hunk 9]
```diff
@@ -38,6 +38,7 @@ master_enc_key = "73ad7bbbbc640c845a150f67d058b279849370cd2c1f3c67c4dd6c869213e1
 host = ""
 mock_locker = true
 basilisk_host = ""
+locker_setup = "legacy_locker"
```
[the remark this answers is not in the document]
Comment (sahkal) [reply config/development.toml:41]: will copy paste development.toml file from main. to avoid sync failure

File: crates/router/src/routes/payment_methods.rs (hunk 7 above)
Comment (sahkal) [reply crates/router/src/routes/payment_methods.rs:158]: sure, will change this

File: crates/router/src/core/payment_methods/cards.rs (hunk 6 above)
Comment (Narayanbhat166) [reply crates/router/src/core/payment_methods/cards.rs:1514]: We need not remove the struct, but the fields can be removed.

File: crates/router/src/compatibility/stripe/customers.rs [hunk 10]
```diff
@@ -168,10 +168,8 @@ pub async fn list_customer_payment_method_api(
     state: web::Data<routes::AppState>,
     req: HttpRequest,
     path: web::Path<String>,
-    json_payload: web::Query<payment_methods::PaymentMethodListRequest>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/customers.rs:171]: Since there are 2 api routes now in core. Can this changes be undo. Make sure it is stripe compatible. https://stripe.com/docs/api/payment_methods/customer_list

File: config/development.toml [hunk 11]
```diff
@@ -132,6 +132,7 @@ globepay.base_url = "https://pay.globepay.co/"
 iatapay.base_url = "https://sandbox.iata-pay.iata.org/api/v1"
 klarna.base_url = "https://api-na.playground.klarna.com/"
 mollie.base_url = "https://api.mollie.com/v2/"
+mollie.secondary_base_url = "https://api.cc.mollie.com/v1/"
```
Comment (Abhicodes-crypto) [inline config/development.toml:135]: Maybe fix this .

File: crates/router/src/compatibility/stripe/customers.rs (hunk 10 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/customers.rs:171]: done

File: openapi/openapi_spec.json [hunk 12]
```diff
@@ -410,6 +416,115 @@
         ]
       }
     },
+    "/customers/{customer_id}/payment_methods": {
```
Comment (ashokkjag) [inline openapi/openapi_spec.json:419]: Both this endpoint and `"/customers/payment_methods" are saying that "list payment methods for a customer, which seems a little confusing. 
Though I would prefer things to be explicit, I'm assuming some need for returning the current customers payment methods exist. So better to clearly label that in the openapi spec/api docs. Clearly mention `/customers/payment_method` returns for which customer(`current?`) and if so how is the customer identified.

Comment (sahkal) [reply openapi/openapi_spec.json:419]: /customers/payment_method, will return paymentMethod for customer that is tied to its client-secret passed in the params.

**Review by ashokkjag — DISMISSED** [review]

File: crates/router/src/core/payment_methods/cards.rs [hunk 13]
```diff
@@ -1556,15 +1588,10 @@ pub async fn list_customer_payment_method(
...
-        let payment_intent = helpers::verify_payment_intent_time_and_client_secret(
-            db,
-            &merchant_account,
-            req.client_secret.clone(),
-        )
-        .await?;
         let current_datetime_utc = common_utils::date_time::now();
         let time_eslapsed = current_datetime_utc
             - payment_intent
+                .clone()
```
Comment (Abhicodes-crypto) [inline crates/router/src/core/payment_methods/cards.rs:1594]:
```suggestion
                .as_ref()
```

File: crates/router/src/core/payment_methods/cards.rs [hunk 14]
```diff
@@ -1499,15 +1499,47 @@ async fn filter_payment_mandate_based(
...
+        let payment_intent = helpers::verify_payment_intent_time_and_client_secret(
+            db,
+            &merchant_account,
+            cloned_secret,
+        )
+        .await?;
+        let customer_id = payment_intent
+            .clone()
+            .and_then(|intent| (intent.customer_id))
+            .ok_or(errors::ApiErrorResponse::CustomerNotFound)?;
```
Comment (Abhicodes-crypto) [inline crates/router/src/core/payment_methods/cards.rs:1523]:
```suggestion
       let customer_id = payment_intent
            .as_ref()
            .and_then(|intent| intent.customer_id.to_owned())
            .ok_or(errors::ApiErrorResponse::CustomerNotFound)?;
```

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs (hunk 13 above)
Comment (sahkal) [reply crates/router/src/core/payment_methods/cards.rs:1594]: done

File: crates/router/src/core/payment_methods/cards.rs (hunk 14 above)
Comment (sahkal) [reply crates/router/src/core/payment_methods/cards.rs:1523]: done

**Review by Abhicodes-crypto — APPROVED** [review]

**Review by Abhicodes-crypto — DISMISSED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]

**Review by ashokkjag — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
